doc/dev/error-categorization.md

```diff
@@ -0,0 +1,167 @@
+`kdbGet`. Also violations of the backend belong into this category. Compared to semantic validation errors,
+this category has its focus on detecting wrong usages of the API instead of a "retry with a different value" approach.
+
+##### Plugin is Broken ("C01330")
+
+`Broken Plugin Errors` errors indicate that a plugin does not behave in an intended way. Unrecognized commands,
+unkown return codes, plugin creation errors, etc. belong to this category. Also uncaught exceptions belong here because
+Elektra expects all exceptions to be caught.
+
+### Conflicting State ("C02000")
+
+`Conflicting State Errors` are errors where the current state is incompatible with the attempted operation.
+Examples are the need for calling `kdbGet` before `kdbSet`. Another example would be to try to push you changes
+into a git repository where the remote branch has already changed.
+These kind of errors are usually in resolver plugins when the state of the file
+has changed without the system knowing. Such errors demand a special synchronization action before retrying.
+
+### Timeout ("C03000")
+
+`Timeout Errors` are easy to categorize as the reaction simply suggests to wait for a short period and retry again.
+Examples are lost connections to a server.
+
+### Validation ("C04000")
+
+`Validation Errors` are heavily used for Elektra's `configuration specification` feature and
+should tell users that their given input does not match a certain pattern/type/expected semantic etc.
+
+Validation errors can either be syntactic or semantic.
+
+#### Syntactic ("C04100")
+
+`Syntactic Errors` are errors which tell users or applications that the current format is not valid.
+Examples are wrong date formats or missing closing brackets `]` inside of a regular expression. Also path related errors
+like missing slashes come into this category. Also parsing errors fall under syntactic errors such as
+unexpected encounters like missing line endings, illegal characters or wrong encodings. Also transformation
+and conversion errors are to be categorizes here because the format of the given input does not allow such
+actions. Users should try a different format and retry pushing it to Elektra.
+
+#### Semantic ("C04200")
+
+`Semantic Errors` are a bit more tricky to solve programmatically. Examples are references to
+non-existent keys (`reference` plugin), setting values to keys which require to be empty (`required` plugin),
+wrong provided values in a specification if you restricted the values (`enum` plugin), etc.
+
+The provided value differs from the developers/ administrators (specification writer) intention,
+even though syntactically everything is valid. Users should try a different value and retry again.
+
+### Out of Range ("C05000")
+
+`Out of Range Errors` are operations which attempt to be past a valid range. Examples are trying to set
+an Elektra Array entry to `-1` or trying to set a port outside of a valid range. There is quite some overlap
```

Comment:
"There is quite some overlap" then we should not have the error. Try to properly distinguish.

If it is only about Elektra's arrays, Validation/Out of Range would be better?

Or we create Validation/Structure for problems in the structure of the KeySet?

@sanssecours Can you maybe help out?

Reply:
> Can you maybe help out?

Nope, sorry. I do not know anything about error categorization and I am also currently not interested in learning more about the topic.

Reply:
Then I think we should remove this category.

Validation/Structure might be useful to pursuit, though.

Reply:
I will merge it back to validation. It was initially here because some errors suggested by sanssecours were difficult to get right at first. This category would be a very clear and intuitive category despite some overlap. 

Such overlaps are btw now uncommon in other error frameworks too such as [gRPC](https://github.com/grpc/grpc/blob/master/doc/statuscodes.md)